clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java

```diff
@@ -284,8 +282,8 @@ public void seek(TopicPartition tp, long offset) {
         return this.assignment.partitionSet();
     }
 
-    public List<TopicPartition> fetchablePartitions() {
-        List<TopicPartition> fetchable = new ArrayList<>();
+    public Set<TopicPartition> fetchablePartitions() {
+        Set<TopicPartition> fetchable = new HashSet<>();
```

Comment:
Order must be maintained, so this changes the behaviour. You'd have to use a `LinkedHashSet`, which is a lot more expensive to create than an `ArrayList` and would probably defeat the point of the change. If we think the usage of a predicate is too awkward, we could replace the two `remove` calls with a single `removeAll`. That would probably be a little cheaper.

Also, we can probably pre-size the `ArrayList` in this method with `assignment.size()`. `isFetchable` should be true in most cases.